### PR TITLE
operator: Add watch on the Alertmanager in OCP's user-workload-monitoring namespace

### PR DESCRIPTION
**Pull request by aminesnow**
**What this PR does / why we need it**:
Following #7839, this PR adds a watch on the Alertmanager in the user-workload-monitoring namespace that is used when the operator is running on an OpenShift cluster. 

**Special notes for your reviewer**:
@periklis let me know if this is what you had in mind.

**Checklist**
- [x] Reviewed the `CONTRIBUTING.md` guide
- [ ] Documentation added
- [ ] Tests updated
- [x] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: operator/controllers/loki/lokistack_controller.go [hunk 1]
```diff
@@ -178,7 +187,8 @@ func (r *LokiStackReconciler) buildController(bld k8s.Builder) error {
 		Owns(&rbacv1.ClusterRole{}, updateOrDeleteOnlyPred).
 		Owns(&rbacv1.ClusterRoleBinding{}, updateOrDeleteOnlyPred).
 		Owns(&rbacv1.Role{}, updateOrDeleteOnlyPred).
-		Owns(&rbacv1.RoleBinding{}, updateOrDeleteOnlyPred)
+		Owns(&rbacv1.RoleBinding{}, updateOrDeleteOnlyPred).
+		Watches(&source.Kind{Type: &corev1.Service{}}, r.enqueueAllLokiStacksHandler(), createUpdateOrDeletePred)
```
Comment (periklis) [inline operator/controllers/loki/lokistack_controller.go:191]: :scream: Why is this possible at all? Using the same source Kind for owned and watched resources?

TL;DR; Although the controller runtime API seems to allow this I suggest to dive into the code behind this to see what happens with caching, listers and informers when you register the same source Kind twice. This seems odd and dangerous.

Besides this the `Watches` clause seems to be also very eager sort of that any service across the cluster will trigger a reconciliation loop for all LokiStack instances. Can't we limit this to the two particular OCP services and only for the LokiStacks with tenant modes `openshift-logging` and `openshift-network`?

**Review by periklis — CHANGES_REQUESTED** [review]

File: operator/controllers/loki/lokistack_controller.go (hunk 1 above)
Comment (aminesnow) [reply operator/controllers/loki/lokistack_controller.go:191]: I agree that it's reckless to have all the SVCs trigger the reconciliation when we are interested in only one specific SVC, so I'll edit this to be more specific. 

However, I don't see an issue with having an Owns and a Watches on the same resource; Owns is just a Watches under the hood, and if I'm not mistaken, we can have multiple Watches on the same resource in order to react to different events differently. If it's not possible then please enlighten me as I might have misunderstood the behavior.

File: operator/controllers/loki/lokistack_controller.go [hunk 2]
```diff
@@ -224,3 +235,31 @@ func (r *LokiStackReconciler) enqueueAllLokiStacksHandler() handler.EventHandler
...
+		if obj.GetName() == openshift.MonitoringSVCOperated && obj.GetNamespace() == openshift.MonitoringUserwWrkloadNS {
+
+			for _, stack := range lokiStacks.Items {
+				requests = append(requests, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: stack.Namespace,
+						Name:      stack.Name,
+					},
+				})
+			}
```
Comment (periklis) [inline operator/controllers/loki/lokistack_controller.go:258]: I suggest to omit any stack that has tenant mode other than `openshift-logging` or `openshift-network`. These are the only two tenants we reconcile OpenShift related Alertmanager configuration.

File: operator/controllers/loki/lokistack_controller.go [hunk 3]
```diff
@@ -224,3 +235,31 @@ func (r *LokiStackReconciler) enqueueAllLokiStacksHandler() handler.EventHandler
...
+	ctx := context.TODO()
+	return handler.EnqueueRequestsFromMapFunc(func(obj client.Object) []reconcile.Request {
+		lokiStacks := &lokiv1.LokiStackList{}
+		if err := r.Client.List(ctx, lokiStacks); err != nil {
+			r.Log.Error(err, "Error getting LokiStack resources in event handler")
+			return nil
+		}
+		var requests []reconcile.Request
+
+		if obj.GetName() == openshift.MonitoringSVCOperated && obj.GetNamespace() == openshift.MonitoringUserwWrkloadNS {
```
Comment (periklis) [inline operator/controllers/loki/lokistack_controller.go:249]: What about the AM in the `openshift-monitoring` namespace?

**Review by periklis — CHANGES_REQUESTED** [review]

File: operator/controllers/loki/lokistack_controller.go (hunk 3 above)
Comment (aminesnow) [reply operator/controllers/loki/lokistack_controller.go:249]: what about it?

Comment (periklis) [reply operator/controllers/loki/lokistack_controller.go:249]: Don't we need to check for it here?

Comment (aminesnow) [reply operator/controllers/loki/lokistack_controller.go:249]: maybe, maybe not, but this is not the PR for that

File: operator/controllers/loki/lokistack_controller.go (hunk 2 above)
Comment (aminesnow) [reply operator/controllers/loki/lokistack_controller.go]: done

File: operator/controllers/loki/lokistack_controller.go (hunk 3 above)
Comment (periklis) [reply operator/controllers/loki/lokistack_controller.go:249]: Why not?

Comment (aminesnow) [reply operator/controllers/loki/lokistack_controller.go:249]: I'd rather get this one right and merged then do another PR for the other AM.

Comment (periklis) [reply operator/controllers/loki/lokistack_controller.go:249]: I see this being the same for both AM as both services have the name `alertmanager-operated` but differ only in a namespace.

Comment (periklis) [reply operator/controllers/loki/lokistack_controller.go:249]: :shrug:

File: operator/controllers/loki/lokistack_controller.go [hunk 4]
```diff
@@ -224,3 +235,34 @@ func (r *LokiStackReconciler) enqueueAllLokiStacksHandler() handler.EventHandler
...
+			for _, stack := range lokiStacks.Items {
+				if stack.Spec.Tenants.Mode == lokiv1.OpenshiftLogging ||
+					stack.Spec.Tenants.Mode == lokiv1.OpenshiftNetwork {
+					requests = append(requests, reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Namespace: stack.Namespace,
+							Name:      stack.Name,
+						},
+					})
+				}
```
Comment (periklis) [inline operator/controllers/loki/lokistack_controller.go:260]: Need to check here if  `stack.Spec.Tenants` is not nil.

**Review by periklis — CHANGES_REQUESTED** [review]

File: operator/controllers/loki/lokistack_controller.go (hunk 4 above)
Comment (aminesnow) [reply operator/controllers/loki/lokistack_controller.go:260]: good catch, thnx!

File: operator/CHANGELOG.md [hunk 5]
```diff
@@ -1,10 +1,11 @@
 ## Main
 
-- [8173](https://github.com/grafana/loki/pull/8173) **periklis**: Remove custom webhook cert mounts for OLM-based deployment (OpenShift)
 - [8001](https://github.com/grafana/loki/pull/8001) **aminesnow**: Add API validation to Alertmanager header auth config
+- [8173](https://github.com/grafana/loki/pull/8173) **periklis**: Remove custom webhook cert mounts for OLM-based deployment (OpenShift)
 - [8087](https://github.com/grafana/loki/pull/8087) **xperimental**: Fix status not updating when state of pods changes
 - [8068](https://github.com/grafana/loki/pull/8068) **periklis**: Use lokistack-gateway replicas from size table
 - [8068](https://github.com/grafana/loki/pull/8068) **periklis**: Use lokistack-gateway replicas from size table
+- [8038](https://github.com/grafana/loki/pull/8038) **aminesnow**: Add watch on the Alertmanager in OCP's user-workload-monitoring namespace
```
Comment (periklis) [inline operator/CHANGELOG.md:8]: You pulled in some entries through the last rebase. In addition Line8 should be on the top.

**Review by periklis — APPROVED** [review]

File: operator/CHANGELOG.md (hunk 5 above)
Comment (aminesnow) [reply operator/CHANGELOG.md:8]: Got confused and moved the wrong line to the top :sweat_smile:

**Review by periklis — APPROVED** [review]
